macros/models/_sector/dex/enrich_dex_trades.sql

```diff
@@ -0,0 +1,83 @@
+{% macro enrich_dex_trades(
+    stg_trades = null
+    , tokens_erc20_model = null
+    , prices_model = null
+    )
+%}
+
+WITH prices AS (
+    SELECT
+        blockchain
+        , contract_address
+        , minute
+        , price
+    FROM
+        {{ prices_model }}
+    {% if is_incremental() %}
+    WHERE
+        {{ incremental_predicate('minute') }}
+    {% endif %}
+)
+, enrichments AS (
+    SELECT
+        base.blockchain
+        , base.project
+        , base.version
+        , base.block_month
+        , base.block_date
+        , base.block_time
+        , base.block_number
+        , erc20_bought.symbol AS token_bought_symbol
+        , erc20_sold.symbol AS token_sold_symbol
+        , case
+            when lower(erc20_bought.symbol) > lower(erc20_sold.symbol) then concat(erc20_sold.symbol, '-', erc20_bought.symbol)
+            else concat(erc20_bought.symbol, '-', erc20_sold.symbol)
+            end AS token_pair
+        , base.token_bought_amount_raw / power(10, erc20_bought.decimals) AS token_bought_amount
+        , base.token_sold_amount_raw / power(10, erc20_sold.decimals) AS token_sold_amount
+        , base.token_bought_amount_raw
+        , base.token_sold_amount_raw
+        , coalesce(
+                base.token_bought_amount_raw / power(10, erc20_bought.decimals) * p_bought.price,
+                base.token_sold_amount_raw / power(10, erc20_sold.decimals) * p_sold.price
+            ) AS amount_usd
+        , base.token_bought_address
+        , base.token_sold_address
+        , coalesce(base.taker, base.tx_from) AS taker
+        , base.maker
+        , base.project_contract_address
+        , base.tx_hash
+        , base.tx_from
+        , base.tx_to
+        , base.evt_index
+    FROM
+        {{ stg_trades }} base
+    LEFT JOIN
+        {{ tokens_erc20_model }} erc20_bought
+        ON erc20_bought.contract_address = base.token_bought_address
+        AND erc20_bought.blockchain = base.blockchain
+    LEFT JOIN
+        {{ tokens_erc20_model }} erc20_sold
+        ON erc20_sold.contract_address = base.token_sold_address
+        AND erc20_sold.blockchain = base.blockchain
+    LEFT JOIN
+        prices as p_bought
+        ON p_bought.minute = date_trunc('minute', base.block_time)
+        AND p_bought.contract_address = base.token_bought_address
+        AND p_bought.blockchain = base.blockchain
+    LEFT JOIN
+        prices as p_sold
+        ON p_sold.minute = date_trunc('minute', base.block_time)
+        AND p_sold.contract_address = base.token_sold_address
+        AND p_sold.blockchain = base.blockchain
+    {% if is_incremental() %}
+    WHERE
+        {{ incremental_predicate('base.block_time') }}
```

Comment:
if you'd move this to the upper level it should improve incremental performance more. i mean:

```
with 

...

stg_cte as (
	select * from {{ stg_trades }}
	{% if is_incremental() %}
		where {{ incremental_predicate('base.block_time') }}
	{% endif %}
)

...

from stg_cte
left join 

...
```

filtering data as early as possible before joins almost always helps.

P.S: also — `base` :/

Reply:
i can give it a shot, but not sure we'll gain much from what i've seen before.

ugh, why did they have to name their blockchain `base`, that's such a useful word for aliases and such 🥲

Reply:
i hope Stargate won't release their STG chain haha

Reply:
> not sure we'll gain much from what i've seen before

`dex_trades_beta  [SUCCESS in 14.06s]`→ `dex_trades_beta  [SUCCESS in 9.90s]`
not much, but still around 1.5x which can easily become x2-x3 on a larger dataset i think

Reply:
@jeff-dude yep it's just bias. Trino's optimizer is pretty smart. 

<img width="502" alt="image" src="https://github.com/duneanalytics/spellbook/assets/40689054/0b9335bb-f703-4275-88d9-673b8b334096">
<img width="1285" alt="image" src="https://github.com/duneanalytics/spellbook/assets/40689054/b08a1d7e-b66a-4d76-be08-3446554a4a9c">

Anyway good to know!

Reply:
Wow which tool are you using here, seems pretty helpful.